### PR TITLE
NIAD-2455  Non functional test - Create and setup docker tests

### PR DESCRIPTION
**Pull request by MartinWheelerMT**

### CODE REVIEW

File: test-suite/non-functional-tests/test-scenario-1/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+12 Transfers requested - 10 expected to succeed, 2 expected to fail, each successful transfer contains:
+* 1 RCMR_IN030000UK06 message
+* 1 COPC_IN000001UK01 index message
+* 2 COPC_IN000001UK01 Fragment messages
+
+### Prerequisites 
+
+You will need to provide the certificate and key required to connect to the inbound service.
+* Copy the certificate and key files into the /testPlans/certs folder named 'client.crt' and 'client.key' respectively.
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/README.md:15]: I think the path here is wrong, and in the next 5 occurances of this sentence too 🤣

File: test-suite/non-functional-tests/test-scenario-1/dockerfile [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+RUN sed -i "s|{{facade-port}}|${FACADE_PORT}|g" ./jmeter-config/config.properties  \
+    && sed -i "s|{{facade-url}}|${FACADE_URL}|g" ./jmeter-config/config.properties \
+    && sed -i "s|{{inbound-port}}|${INBOUND_PORT}|g" ./jmeter-config/config.properties \
+    && sed -i "s|{{inbound-url}}|${INBOUND_URL}|g" ./jmeter-config/config.properties \
+    && sed -i "s|{{jms-provider-path}}|${JNDI_PROPERTIES_PATH}|g" ./jmeter-config/config.properties
+
+COPY testplan.jmx ./jmeter-config/
+COPY run-tests.sh ./jmeter-config/
+COPY jmeterkeystore.p12 ./jmeter-config/certs/jmeterkeystore.p12
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/dockerfile:50]: I general, I'd be reluctant to store the client certs within the docker image, but given this is a test image not that important.

File: test-suite/non-functional-tests/test-scenario-1/testplan.jmx
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/testplan.jmx:1]: If you'd like me to review this file, you're going to have to help show me how to review `jmx`.

File: test-suite/non-functional-tests/test-scenario-1/dockerfile [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ARG FACADE_URL 
+ARG FACADE_PORT 
+ARG INBOUND_URL
+ARG INBOUND_PORT
+
+RUN apk update  \
+    && apk upgrade  \
+    && apk add --update openjdk8-jre tzdata curl unzip bash  \
+    && rm -rf /var/cache/apk/*  \
+    && mkdir -p /tmp/dependencies/ \
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/dockerfile:22]: Minor feedback, `mkdir` can accept multiple directories.

Also `-p` will create intermediate directories, so `mkdir -p ./jmeter-config/certs` will create `./jmeter-config` if it doesn't already exist.

`mkdir -p /tmp/dependeices/ /opt ./jmeter-logs ./jmeter-config/certs`

File: test-suite/non-functional-tests/test-scenario-1/dockerfile [hunk 4]
```diff
@@ -0,0 +1,56 @@
+FROM openjdk:17-jdk-alpine
+
+ARG JMETER_VERSION="5.5"
+ARG JMETER_HOME="/opt/apache-jmeter-${JMETER_VERSION}"
+ARG JMETER_BIN="${JMETER_HOME}/bin"
+ARG MIRROR_HOST=http://mirrors.ocf.berkeley.edu/apache/jmeter
+ARG JMETER_DOWNLOAD_URL="${MIRROR_HOST}/binaries/apache-jmeter-${JMETER_VERSION}.tgz"
+ARG JMETER_PLUGINS_DOWNLOAD_URL="http://repo1.maven.org/maven2/kg/apc"
+ARG JMETER_LIB_FOLDER="${JMETER_HOME}/lib/"
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/dockerfile:9]: Don't think this variable is being used.

File: test-suite/non-functional-tests/test-scenario-1/dockerfile [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ARG JMETER_PLUGINS_FOLDER="${JMETER_HOME}/lib/ext/"
+ARG INBOUND_CERT_PATH="/jmeter-config/certs"
+
+ARG FACADE_URL 
+ARG FACADE_PORT 
+ARG INBOUND_URL
+ARG INBOUND_PORT
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/dockerfile:16]: Minor, I'd have made these environment variables, thus preveting the need for a rebuild each time you wanted to test against a different host.

File: test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+openssl pkcs12 -export -out jmeterkeystore.p12 -inkey ../certs/client.key -in ../certs/client.crt -passout pass:password
+
+echo Stopping docker container:
+docker ps -a -q -f name=nf-test-1 | xargs -r docker stop
+echo Removing Docker container:
+docker ps -a -q -f name=nf-test-1 | xargs -r docker rm
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh:8]: Would `docker compose rm -s -f` be simpler?

File: test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+openssl pkcs12 -export -out jmeterkeystore.p12 -inkey ../certs/client.key -in ../certs/client.crt -passout pass:password
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh:3]: Is this a hardcoded passphrase?

Comment (adrianclay) [reply test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh:3]: I'm _guessing_ there is some technical need to password protect the certificate? Seems annoying.

File: test-suite/non-functional-tests/test-scenario-1/README.md (hunk 1 above)
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/README.md]: Corrected in all places

File: test-suite/non-functional-tests/test-scenario-1/dockerfile (hunk 3 above)
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/dockerfile]: ahh was not aware of that, have simplified in all the files

File: test-suite/non-functional-tests/test-scenario-1/dockerfile [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+ARG JMETER_VERSION="5.5"
+ARG JMETER_HOME="/opt/apache-jmeter-${JMETER_VERSION}"
+ARG JMETER_BIN="${JMETER_HOME}/bin"
+ARG MIRROR_HOST=http://mirrors.ocf.berkeley.edu/apache/jmeter
+ARG JMETER_DOWNLOAD_URL="${MIRROR_HOST}/binaries/apache-jmeter-${JMETER_VERSION}.tgz"
+ARG JMETER_PLUGINS_DOWNLOAD_URL="http://repo1.maven.org/maven2/kg/apc"
+ARG JMETER_LIB_FOLDER="${JMETER_HOME}/lib/"
+ARG JMETER_PLUGINS_FOLDER="${JMETER_HOME}/lib/ext/"
+ARG INBOUND_CERT_PATH="/jmeter-config/certs"
```
[the remark this answers is not in the document]
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/dockerfile]: removed this variable as no longer needed

File: test-suite/non-functional-tests/test-scenario-1/dockerfile (hunk 4 above)
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/dockerfile]: removed this variable as no longer needed

File: test-suite/non-functional-tests/test-scenario-1/dockerfile (hunk 5 above)
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/dockerfile]: completed these changes, makes sense the way you have explained

File: test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh (hunk 6 above)
Comment (MartinWheelerMT) [reply test-suite/non-functional-tests/test-scenario-1/start-docker-test.sh]: yep, tested that and that works so have updated to this

File: test-suite/non-functional-tests/test-scenario-1/testplan.jmx
Comment (adrianclay) [reply test-suite/non-functional-tests/test-scenario-1/testplan.jmx:1]: Well so far I've established that we seem to be sending war and peace, and canterbury tales. I'm a little suspicious of the licencing implications of including these in our repo.

**Review by adrianclay — APPROVED** [review]

**Pull request merged**